File: internal/provider/passthrough.go
package provider

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"github.com/tjfontaine/polyglot-llm-gateway/internal/domain"
)

// PassthroughProvider wraps a provider and bypasses canonical conversion
// when the source API type matches the provider's API type.
type PassthroughProvider struct {
	inner      domain.Provider
	apiKey     string
	baseURL    string
	httpClient *http.Client
}

// PassthroughOption configures the passthrough provider.
type PassthroughOption func(*PassthroughProvider)

// WithPassthroughAPIKey sets the API key for pass-through requests.
func WithPassthroughAPIKey(apiKey string) PassthroughOption {
	return func(p *PassthroughProvider) {
		p.apiKey = apiKey
	}
}

// WithPassthroughBaseURL sets the base URL for pass-through requests.
func WithPassthroughBaseURL(baseURL string) PassthroughOption {
	return func(p *PassthroughProvider) {
		p.baseURL = baseURL
	}
}

// WithPassthroughHTTPClient sets the HTTP client for pass-through requests.
func WithPassthroughHTTPClient(client *http.Client) PassthroughOption {
	return func(p *PassthroughProvider) {
		p.httpClient = client
	}
}

// NewPassthroughProvider creates a new pass-through provider wrapper.
func NewPassthroughProvider(inner domain.Provider, opts ...PassthroughOption) *PassthroughProvider {
	p := &PassthroughProvider{
		inner:      inner,
		httpClient: http.DefaultClient,
	}

	for _, opt := range opts {
		opt(p)
	}

	// Set default base URLs based on provider type
	if p.baseURL == "" {
		switch inner.APIType() {
		case domain.APITypeOpenAI:
			p.baseURL = "https://api.openai.com/v1"
		case domain.APITypeAnthropic:
			p.baseURL = "https://api.anthropic.com/v1"
		}
	}

	return p
}

func (p *PassthroughProvider) Name() string {
	return p.inner.Name()
}

func (p *PassthroughProvider) APIType() domain.APIType {
	return p.inner.APIType()
}

// SupportsPassthrough returns true if the provider can handle raw requests
// from the given API type.
func (p *PassthroughProvider) SupportsPassthrough(sourceType domain.APIType) bool {
	return sourceType == p.inner.APIType() && len(p.apiKey) > 0
}

// Complete handles requests, using pass-through when possible.
func (p *PassthroughProvider) Complete(ctx context.Context, req *domain.CanonicalRequest) (*domain.CanonicalResponse, error) {
	// Check if we can use pass-through
	if p.SupportsPassthrough(req.SourceAPIType) && len(req.RawRequest) > 0 {
		rawResp, err := p.CompleteRaw(ctx, req)
		if err != nil {
			return nil, err
		}

		// Parse raw response into canonical format for recording
		resp, err := p.parseRawResponse(rawResp)
		if err != nil {
			return nil, err
		}
		resp.RawResponse = rawResp
		resp.SourceAPIType = p.inner.APIType()
		return resp, nil
	}

	// Fall back to canonical conversion
	return p.inner.Complete(ctx, req)
}

// CompleteRaw handles a raw request body and returns a raw response body.
func (p *PassthroughProvider) CompleteRaw(ctx context.Context, req *domain.CanonicalRequest) ([]byte, error) {
	var endpoint string
	var headers map[string]string

	switch p.inner.APIType() {
	case domain.APITypeOpenAI:
		endpoint = p.baseURL + "/chat/completions"
		headers = map[string]string{
			"Authorization": "Bearer " + p.apiKey,
			"Content-Type":  "application/json",
		}
	case domain.APITypeAnthropic:
		endpoint = p.baseURL + "/messages"
		headers = map[string]string{
			"x-api-key":         p.apiKey,
			"anthropic-version": "2023-06-01",
			"Content-Type":      "application/json",
		}
	default:
		return nil, fmt.Errorf("unsupported API type for passthrough: %s", p.inner.APIType())
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(req.RawRequest))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	for k, v := range headers {
		httpReq.Header.Set(k, v)
	}

	if req.UserAgent != "" {
		httpReq.Header.Set("User-Agent", req.UserAgent)
	}

	resp, err := p.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API error (status %d): %s", resp.StatusCode, string(body))
	}

	return body, nil
}

// Stream handles streaming requests, using pass-through when possible.
func (p *PassthroughProvider) Stream(ctx context.Context, req *domain.CanonicalRequest) (<-chan domain.CanonicalEvent, error) {
	// Check if we can use pass-through
	if p.SupportsPassthrough(req.SourceAPIType) && len(req.RawRequest) > 0 {
		return p.StreamRaw(ctx, req)
	}

	// Fall back to canonical conversion
	return p.inner.Stream(ctx, req)
}

// StreamRaw handles a raw streaming request and returns canonical events with raw data attached.
func (p *PassthroughProvider) StreamRaw(ctx context.Context, req *domain.CanonicalRequest) (<-chan domain.CanonicalEvent, error) {
	var endpoint string
	var headers map[string]string

	switch p.inner.APIType() {
	case domain.APITypeOpenAI:
		endpoint = p.baseURL + "/chat/completions"
		headers = map[string]string{
			"Authorization": "Bearer " + p.apiKey,
			"Content-Type":  "application/json",
		}
	case domain.APITypeAnthropic:
		endpoint = p.baseURL + "/messages"
		headers = map[string]string{
			"x-api-key":         p.apiKey,
			"anthropic-version": "2023-06-01",
			"Content-Type":      "application/json",
		}
	default:
		return nil, fmt.Errorf("unsupported API type for passthrough: %s", p.inner.APIType())
	}

	// Modify raw request to ensure streaming is enabled
	var rawReq map[string]interface{}
	if err := json.Unmarshal(req.RawRequest, &rawReq); err != nil {
		return nil, fmt.Errorf("failed to parse raw request: %w", err)
	}
	rawReq["stream"] = true
	modifiedReq, err := json.Marshal(rawReq)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal modified request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(modifiedReq))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	for k, v := range headers {
		httpReq.Header.Set(k, v)
	}

	if req.UserAgent != "" {
		httpReq.Header.Set("User-Agent", req.UserAgent)
	}

	resp, err := p.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		defer resp.Body.Close()
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("API error (status %d): %s", resp.StatusCode, string(body))
	}

	// Return a channel that wraps the raw SSE stream
	out := make(chan domain.CanonicalEvent)
	go p.streamRawReader(resp.Body, out, p.inner.APIType())
	return out, nil
}

func (p *PassthroughProvider) streamRawReader(body io.ReadCloser, out chan<- domain.CanonicalEvent, apiType domain.APIType) {
	defer close(out)
	defer body.Close()

	buf := make([]byte, 4096)
	var eventType string
	var data []byte

	for {
		n, err := body.Read(buf)
		if n > 0 {
			// Parse SSE format
			lines := bytes.Split(buf[:n], []byte("\n"))
			for _, line := range lines {
				line = bytes.TrimSpace(line)
				if len(line) == 0 {
					// Empty line means end of event
					if len(data) > 0 {
						event := p.parseStreamEvent(eventType, data, apiType)
						if event.Error != nil || event.ContentDelta != "" || event.Usage != nil {
							event.RawEvent = data
							out <- event
						}
						data = nil
						eventType = ""
					}
					continue
				}

				if bytes.HasPrefix(line, []byte("event:")) {
					eventType = string(bytes.TrimSpace(bytes.TrimPrefix(line, []byte("event:"))))
				} else if bytes.HasPrefix(line, []byte("data:")) {
					d := bytes.TrimSpace(bytes.TrimPrefix(line, []byte("data:")))
					if string(d) == "[DONE]" {
						return
					}
					data = d
				}
			}
		}

		if err != nil {
			if err != io.EOF {
				out <- domain.CanonicalEvent{Error: fmt.Errorf("stream read error: %w", err)}
			}
			return
		}
	}
}

func (p *PassthroughProvider) parseStreamEvent(eventType string, data []byte, apiType domain.APIType) domain.CanonicalEvent {
	event := domain.CanonicalEvent{}

	switch apiType {
	case domain.APITypeOpenAI:
		var chunk struct {
			Choices []struct {
				Delta struct {
					Content string `json:"content"`
					Role    string `json:"role"`
				} `json:"delta"`
			} `json:"choices"`
			Usage *struct {
				PromptTokens     int `json:"prompt_tokens"`
				CompletionTokens int `json:"completion_tokens"`
				TotalTokens      int `json:"total_tokens"`
			} `json:"usage"`
		}
		if err := json.Unmarshal(data, &chunk); err != nil {
			event.Error = fmt.Errorf("failed to parse chunk: %w", err)
			return event
		}
		if len(chunk.Choices) > 0 {
			event.ContentDelta = chunk.Choices[0].Delta.Content
			event.Role = chunk.Choices[0].Delta.Role
		}
		if chunk.Usage != nil {
			event.Usage = &domain.Usage{
				PromptTokens:     chunk.Usage.PromptTokens,
				CompletionTokens: chunk.Usage.CompletionTokens,
				TotalTokens:      chunk.Usage.TotalTokens,
			}
		}

	case domain.APITypeAnthropic:
		switch eventType {
		case "content_block_delta":
			var delta struct {
				Delta struct {
					Text string `json:"text"`
				} `json:"delta"`
			}
			if err := json.Unmarshal(data, &delta); err == nil {
				event.ContentDelta = delta.Delta.Text
			}
		case "message_start":
			var start struct {
				Message struct {
					Role  string `json:"role"`
					Usage struct {
						InputTokens int `json:"input_tokens"`
					} `json:"usage"`
				} `json:"message"`
			}
			if err := json.Unmarshal(data, &start); err == nil {
				event.Role = start.Message.Role
				event.Usage = &domain.Usage{
					PromptTokens: start.Message.Usage.InputTokens,
				}
			}
		case "message_delta":
			var delta struct {
				Usage struct {
					OutputTokens int `json:"output_tokens"`
				} `json:"usage"`
			}
			if err := json.Unmarshal(data, &delta); err == nil && delta.Usage.OutputTokens > 0 {
				event.Usage = &domain.Usage{
					CompletionTokens: delta.Usage.OutputTokens,
				}
			}
		}
	}

	return event
}

func (p *PassthroughProvider) parseRawResponse(rawResp []byte) (*domain.CanonicalResponse, error) {
	resp := &domain.CanonicalResponse{}

	switch p.inner.APIType() {
	case domain.APITypeOpenAI:
		var apiResp struct {
			ID      string `json:"id"`
			Object  string `json:"object"`
			Created int64  `json:"created"`
			Model   string `json:"model"`
			Choices []struct {
				Index   int `json:"index"`
				Message struct {
					Role    string `json:"role"`
					Content string `json:"content"`
				} `json:"message"`
				FinishReason string `json:"finish_reason"`
			} `json:"choices"`
			Usage struct {
				PromptTokens     int `json:"prompt_tokens"`
				CompletionTokens int `json:"completion_tokens"`
				TotalTokens      int `json:"total_tokens"`
			} `json:"usage"`
		}
		if err := json.Unmarshal(rawResp, &apiResp); err != nil {
			return nil, fmt.Errorf("failed to parse response: %w", err)
		}

		resp.ID = apiResp.ID
		resp.Object = apiResp.Object
		resp.Created = apiResp.Created
		resp.Model = apiResp.Model
		resp.Usage = domain.Usage{
			PromptTokens:     apiResp.Usage.PromptTokens,
			CompletionTokens: apiResp.Usage.CompletionTokens,
			TotalTokens:      apiResp.Usage.TotalTokens,
		}

		for _, c := range apiResp.Choices {
			resp.Choices = append(resp.Choices, domain.Choice{
				Index: c.Index,
				Message: domain.Message{
					Role:    c.Message.Role,
					Content: c.Message.Content,
				},
				FinishReason: c.FinishReason,
			})
		}

	case domain.APITypeAnthropic:
		var apiResp struct {
			ID      string `json:"id"`
			Type    string `json:"type"`
			Role    string `json:"role"`
			Model   string `json:"model"`
			Content []struct {
				Type string `json:"type"`
				Text string `json:"text"`
			} `json:"content"`
			StopReason string `json:"stop_reason"`
			Usage      struct {
				InputTokens  int `json:"input_tokens"`
				OutputTokens int `json:"output_tokens"`
			} `json:"usage"`
		}
		if err := json.Unmarshal(rawResp, &apiResp); err != nil {
			return nil, fmt.Errorf("failed to parse response: %w", err)
		}

		var content string
		for _, c := range apiResp.Content {
			if c.Type == "text" {
				content += c.Text
			}
		}

		resp.ID = apiResp.ID
		resp.Object = "chat.completion"
		resp.Model = apiResp.Model
		resp.Usage = domain.Usage{
			PromptTokens:     apiResp.Usage.InputTokens,
			CompletionTokens: apiResp.Usage.OutputTokens,
			TotalTokens:      apiResp.Usage.InputTokens + apiResp.Usage.OutputTokens,
		}
		resp.Choices = []domain.Choice{
			{
				Index: 0,
				Message: domain.Message{
					Role:    apiResp.Role,
					Content: content,
				},
				FinishReason: apiResp.StopReason,
			},
		}
	}

	return resp, nil
}

func (p *PassthroughProvider) ListModels(ctx context.Context) (*domain.ModelList, error) {
	return p.inner.ListModels(ctx)
}

// Ensure PassthroughProvider implements the extended interface
var _ domain.PassthroughProvider = (*PassthroughProvider)(nil)

File: ts/packages/gateway-core/src/providers/passthrough.ts
/**
 * Passthrough provider - bypasses canonical conversion when source/target API types match.
 *
 * This is an optimization that sends raw request bytes directly to the provider
 * when the incoming API format matches the provider's API format, avoiding
 * unnecessary encode/decode cycles.
 *
 * @module providers/passthrough
 */

import type {
    CanonicalRequest,
    CanonicalResponse,
    CanonicalEvent,
    APIType,
    Choice,
    Usage,
    Message,
    ModelList,
} from '../domain/types.js';
import type { Provider } from '../ports/provider.js';

// ============================================================================
// Types
// ============================================================================

/**
 * Options for the passthrough provider.
 */
export interface PassthroughOptions {
    /** API key for the provider. */
    apiKey: string;

    /** Base URL for the provider API. */
    baseURL?: string | undefined;

    /** Default headers to include in requests. */
    headers?: Record<string, string> | undefined;
}

/**
 * Extended provider interface with passthrough support.
 */
export interface PassthroughableProvider extends Provider {
    /**
     * Returns true if this provider can handle raw requests from the given API type.
     */
    supportsPassthrough(sourceType: APIType): boolean;

    /**
     * Handles a raw request (when passthrough is supported).
     */
    completeRaw?(request: CanonicalRequest): Promise<[Uint8Array, CanonicalResponse]>;

    /**
     * Handles a raw streaming request.
     */
    streamRaw?(request: CanonicalRequest): AsyncGenerator<CanonicalEvent>;
}

// ============================================================================
// Passthrough Provider
// ============================================================================

/**
 * Wraps a provider to add passthrough support.
 *
 * When the source API type matches the provider's API type and a raw request
 * is available, this provider sends the request directly without canonical
 * conversion, improving latency.
 */
export class PassthroughProvider implements Provider {
    readonly name: string;
    readonly apiType: APIType;
    private readonly inner: Provider;
    private readonly apiKey: string;
    private readonly baseURL: string;
    private readonly headers?: Record<string, string>;

    constructor(inner: Provider, options: PassthroughOptions) {
        this.inner = inner;
        this.name = inner.name;
        this.apiType = inner.apiType;
        this.apiKey = options.apiKey;
        this.baseURL = options.baseURL ?? this.getDefaultBaseURL(inner.apiType);
        this.headers = options.headers;
    }

    /**
     * Returns true if this provider can handle raw requests from the given API type.
     */
    supportsPassthrough(sourceType: APIType): boolean {
        return sourceType === this.apiType && !!this.apiKey;
    }

    /**
     * Completes a request, using passthrough when possible.
     */
    async complete(request: CanonicalRequest): Promise<CanonicalResponse> {
        // Check if we can use passthrough
        if (this.supportsPassthrough(request.sourceAPIType) && request.rawRequest?.length) {
            const [rawResponse, parsedResponse] = await this.completeRaw(request);
            parsedResponse.rawResponse = rawResponse;
            return parsedResponse;
        }

        // Fall back to canonical conversion
        return this.inner.complete(request);
    }

    /**
     * Streams a request, using passthrough when possible.
     */
    stream(request: CanonicalRequest): AsyncGenerator<CanonicalEvent> {
        // Check if we can use passthrough
        if (this.supportsPassthrough(request.sourceAPIType) && request.rawRequest?.length) {
            return this.streamRaw(request);
        }

        // Fall back to canonical conversion
        return this.inner.stream(request);
    }

    /**
     * Lists available models (passes through to inner provider).
     */
    async listModels(): Promise<ModelList> {
        if (this.inner.listModels) {
            return this.inner.listModels();
        }
        return { object: 'list', data: [] };
    }

    // ---- Private Methods ----

    /**
     * Completes a raw request.
     */
    private async completeRaw(request: CanonicalRequest): Promise<[Uint8Array, CanonicalResponse]> {
        const { endpoint, headers } = this.getRequestConfig(this.apiType);

        const response = await fetch(endpoint, {
            method: 'POST',
            headers,
            body: request.rawRequest,
        });

        const rawResponse = new Uint8Array(await response.arrayBuffer());

        if (!response.ok) {
            throw new Error(
                `Provider API error (${response.status}): ${new TextDecoder().decode(rawResponse)}`,
            );
        }

        // Parse the response for canonical format (for recording/logging)
        const parsedResponse = this.parseRawResponse(rawResponse);
        parsedResponse.sourceAPIType = this.apiType;

        return [rawResponse, parsedResponse];
    }

    /**
     * Streams a raw request.
     */
    private async *streamRaw(request: CanonicalRequest): AsyncGenerator<CanonicalEvent> {
        const { endpoint, headers } = this.getRequestConfig(this.apiType);

        // Ensure streaming is enabled in the raw request
        const rawRequest = this.ensureStreamingEnabled(request.rawRequest!);

        const response = await fetch(endpoint, {
            method: 'POST',
            headers,
            body: rawRequest,
        });

        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`Provider API error (${response.status}): ${errorText}`);
        }

        if (!response.body) {
            throw new Error('No response body for streaming');
        }

        // Parse SSE stream
        yield* this.parseSSEStream(response.body, this.apiType);
    }

    /**
     * Gets the default base URL for an API type.
     */
    private getDefaultBaseURL(apiType: APIType): string {
        switch (apiType) {
            case 'openai':
                return 'https://api.openai.com/v1';
            case 'anthropic':
                return 'https://api.anthropic.com/v1';
            default:
                return '';
        }
    }

    /**
     * Gets request configuration for an API type.
     */
    private getRequestConfig(apiType: APIType): { endpoint: string; headers: Record<string, string> } {
        const commonHeaders = {
            'Content-Type': 'application/json',
            ...this.headers,
        };

        switch (apiType) {
            case 'openai':
                return {
                    endpoint: `${this.baseURL}/chat/completions`,
                    headers: {
                        ...commonHeaders,
                        Authorization: `Bearer ${this.apiKey}`,
                    },
                };
            case 'anthropic':
                return {
                    endpoint: `${this.baseURL}/messages`,
                    headers: {
                        ...commonHeaders,
                        'x-api-key': this.apiKey,
                        'anthropic-version': '2023-06-01',
                    },
                };
            default:
                throw new Error(`Unsupported API type for passthrough: ${apiType}`);
        }
    }

    /**
     * Ensures the raw request has streaming enabled.
     */
    private ensureStreamingEnabled(rawRequest: Uint8Array): Uint8Array {
        try {
            const text = new TextDecoder().decode(rawRequest);
            const obj = JSON.parse(text);
            obj.stream = true;
            return new TextEncoder().encode(JSON.stringify(obj));
        } catch {
            return rawRequest;
        }
    }

    /**
     * Parses a raw response into canonical format.
     */
    private parseRawResponse(rawResponse: Uint8Array): CanonicalResponse {
        const text = new TextDecoder().decode(rawResponse);
        const obj = JSON.parse(text);

        switch (this.apiType) {
            case 'openai':
                return this.parseOpenAIResponse(obj);
            case 'anthropic':
                return this.parseAnthropicResponse(obj);
            default:
                throw new Error(`Unsupported API type: ${this.apiType}`);
        }
    }

    /**
     * Parses an OpenAI API response.
     */
    private parseOpenAIResponse(obj: Record<string, unknown>): CanonicalResponse {
        const choices = (obj.choices as Array<{
            index: number;
            message: { role: string; content: string };
            finish_reason: string;
        }>).map((c): Choice => ({
            index: c.index,
            message: {
                role: c.message.role as 'assistant',
                content: c.message.content || '',
            },
            finishReason: c.finish_reason as 'stop' | 'length' | null,
        }));

        const usage = obj.usage as { prompt_tokens: number; completion_tokens: number; total_tokens: number };

        return {
            id: obj.id as string,
            object: obj.object as string,
            created: obj.created as number,
            model: obj.model as string,
            choices,
            usage: {
                promptTokens: usage.prompt_tokens,
                completionTokens: usage.completion_tokens,
                totalTokens: usage.total_tokens,
            },
            sourceAPIType: 'openai',
        };
    }

    /**
     * Parses an Anthropic API response.
     */
    private parseAnthropicResponse(obj: Record<string, unknown>): CanonicalResponse {
        const content = obj.content as Array<{ type: string; text?: string }>;
        const textContent = content
            .filter((c) => c.type === 'text')
            .map((c) => c.text || '')
            .join('');

        const usage = obj.usage as { input_tokens: number; output_tokens: number };

        const message: Message = {
            role: obj.role as 'assistant',
            content: textContent,
        };

        return {
            id: obj.id as string,
            object: 'chat.completion',
            created: Date.now(),
            model: obj.model as string,
            choices: [{
                index: 0,
                message,
                finishReason: this.mapAnthropicStopReason(obj.stop_reason as string),
            }],
            usage: {
                promptTokens: usage.input_tokens,
                completionTokens: usage.output_tokens,
                totalTokens: usage.input_tokens + usage.output_tokens,
            },
            sourceAPIType: 'anthropic',
        };
    }

    /**
     * Maps Anthropic stop reason to canonical finish reason.
     */
    private mapAnthropicStopReason(reason: string): 'stop' | 'length' | 'tool_calls' | null {
        switch (reason) {
            case 'end_turn':
                return 'stop';
            case 'max_tokens':
                return 'length';
            case 'tool_use':
                return 'tool_calls';
            default:
                return null;
        }
    }

    /**
     * Parses an SSE stream into canonical events.
     */
    private async *parseSSEStream(
        body: ReadableStream<Uint8Array>,
        apiType: APIType,
    ): AsyncGenerator<CanonicalEvent> {
        const reader = body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';
        let currentEventType = '';

        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed) {
                        // Empty line means end of event
                        currentEventType = '';
                        continue;
                    }

                    if (trimmed.startsWith('event:')) {
                        currentEventType = trimmed.slice(6).trim();
                        continue;
                    }

                    if (trimmed.startsWith('data:')) {
                        const data = trimmed.slice(5).trim();
                        if (data === '[DONE]') {
                            yield { type: 'done' };
                            return;
                        }

                        try {
                            const parsed = JSON.parse(data);
                            const event = this.parseStreamEvent(parsed, apiType, currentEventType);
                            if (event) {
                                event.rawEvent = new TextEncoder().encode(data);
                                yield event;
                            }
                        } catch {
                            // Skip unparseable events
                        }
                    }
                }
            }
        } finally {
            reader.releaseLock();
        }
    }

    /**
     * Parses a single stream event.
     */
    private parseStreamEvent(
        obj: Record<string, unknown>,
        apiType: APIType,
        eventType: string,
    ): CanonicalEvent | null {
        switch (apiType) {
            case 'openai':
                return this.parseOpenAIStreamEvent(obj);
            case 'anthropic':
                return this.parseAnthropicStreamEvent(obj, eventType);
            default:
                return null;
        }
    }

    /**
     * Parses an OpenAI streaming event.
     */
    private parseOpenAIStreamEvent(obj: Record<string, unknown>): CanonicalEvent | null {
        const choices = obj.choices as Array<{
            delta?: { content?: string; role?: string };
        }> | undefined;

        if (choices?.[0]?.delta?.content) {
            return {
                type: 'content_delta',
                contentDelta: choices[0].delta.content,
            };
        }

        if (choices?.[0]?.delta?.role) {
            return {
                type: 'message_start',
                role: choices[0].delta.role,
            };
        }

        const usage = obj.usage as { prompt_tokens?: number; completion_tokens?: number; total_tokens?: number } | undefined;
        if (usage) {
            return {
                type: 'message_delta',
                usage: {
                    promptTokens: usage.prompt_tokens || 0,
                    completionTokens: usage.completion_tokens || 0,
                    totalTokens: usage.total_tokens || 0,
                },
            };
        }

        return null;
    }

    /**
     * Parses an Anthropic streaming event.
     */
    private parseAnthropicStreamEvent(
        obj: Record<string, unknown>,
        eventType: string,
    ): CanonicalEvent | null {
        switch (eventType) {
            case 'content_block_delta': {
                const delta = obj.delta as { type: string; text?: string } | undefined;
                if (delta?.type === 'text_delta' && delta.text) {
                    return {
                        type: 'content_delta',
                        contentDelta: delta.text,
                    };
                }
                break;
            }
            case 'message_start': {
                const message = obj.message as { role?: string; usage?: { input_tokens?: number } } | undefined;
                if (message) {
                    return {
                        type: 'message_start',
                        role: message.role,
                        usage: message.usage?.input_tokens
                            ? { promptTokens: message.usage.input_tokens, completionTokens: 0, totalTokens: message.usage.input_tokens }
                            : undefined,
                    };
                }
                break;
            }
            case 'message_delta': {
                const usage = obj.usage as { output_tokens?: number } | undefined;
                if (usage?.output_tokens) {
                    return {
                        type: 'message_delta',
                        usage: {
                            promptTokens: 0,
                            completionTokens: usage.output_tokens,
                            totalTokens: usage.output_tokens,
                        },
                    };
                }
                break;
            }
            case 'message_stop':
                return { type: 'message_stop' };
        }

        return null;
    }
}

// ============================================================================
// Factory Function
// ============================================================================

/**
 * Wraps a provider with passthrough support.
 */
export function withPassthrough(
    provider: Provider,
    options: PassthroughOptions,
): PassthroughProvider {
    return new PassthroughProvider(provider, options);
}
